src/odb_backend.c

```diff
@@ -36,6 +36,117 @@
+py_odb_backend_read(void **ptr, size_t *sz,
+        git_object_t *type, git_odb_backend *_be, const git_oid *oid)
+{
+    int err;
+    PyObject *args, *py_oid, *result;
+    struct py_odb_backend *be = (struct py_odb_backend *)_be;
+
+    py_oid = git_oid_to_python(oid);
+    args = Py_BuildValue("(O)", py_oid);
+    result = PyObject_CallObject(be->read_callable, args);
+    Py_DECREF(py_oid);
+
+    if ((err = git_error_for_exc()) != 0) {
+        return err;
+    }
+
+    if (result == NULL)
+        return GIT_EUSER;
+
+    const char *bytes;
+    if (!PyArg_ParseTuple(result, "iy#", type, &bytes, sz) || !bytes)
+        return GIT_EUSER;
+
+    /* XXX: This assumes the default libgit2 allocator is in use and will
+     * probably segfault and/or destroy the universe otherwise */
+    *ptr = malloc(*sz);
+    if (!*ptr)
+        return GIT_EUSER;
+
+    memcpy(*ptr, bytes, *sz);
+    return 0;
+}
+
+static int
+py_odb_backend_read_prefix(git_oid *oid_out, void **ptr, size_t *sz,
+        git_object_t *type, git_odb_backend *_be,
+        const git_oid *short_oid, size_t len)
+{
+    int err;
+    PyObject *args, *py_oid, *py_oid_out, *result;
+    struct py_odb_backend *be = (struct py_odb_backend *)_be;
+
+    py_oid = git_oid_to_python(short_oid);
+    args = Py_BuildValue("(O)", py_oid);
+    result = PyObject_CallObject(be->read_prefix_callable, args);
+    Py_DECREF(py_oid);
+
+    if ((err = git_error_for_exc()) != 0) {
+        return err;
+    }
+
+    if (result == NULL)
+        return GIT_EUSER;
+
+    const char *bytes;
+    if (!PyArg_ParseTuple(result, "Oiy#",
+                &py_oid_out, type, &bytes, sz) || !bytes)
+        return GIT_EUSER;
+
+    /* XXX: This assumes the default libgit2 allocator is in use and will
+     * probably segfault and/or destroy the universe otherwise */
+    *ptr = malloc(*sz);
+    if (!*ptr)
+        return GIT_EUSER;
+
+    memcpy(*ptr, bytes, *sz);
+    py_oid_to_git_oid(py_oid_out, oid_out);
+    return 0;
+}
+
+int
+OdbBackend_init(OdbBackend *self, PyObject *args, PyObject *kwds)
+{
+    if (kwds && PyTuple_Size(args) > 0) {
+        PyErr_SetString(PyExc_TypeError,
+                        "OdbBackend takes no arguments");
+        return -1;
+    }
+
+    if (kwds && PyDict_Size(kwds) > 0) {
+        PyErr_SetString(PyExc_TypeError,
+                        "OdbBackend takes no keyword arguments");
+        return -1;
+    }
+
+    struct py_odb_backend *be = calloc(1, sizeof(struct py_odb_backend));
+    be->backend.version = GIT_ODB_BACKEND_VERSION;
+    be->self = (PyObject *)self;
+
+    be->read_callable = PyObject_GetAttrString((PyObject *)self, "read");
+    if (be->read_callable)
+        be->backend.read = py_odb_backend_read;
+
+    be->read_prefix_callable = PyObject_GetAttrString(
+            (PyObject *)self, "read_prefix");
+    if (be->read_prefix_callable)
+        be->backend.read_prefix = py_odb_backend_read_prefix;
+
+    self->odb_backend = &be->backend;
```

Comment:
I think this would be easier to understand if written:

    self->odb_backend = (git_odb_backend*)be;